File: python/fundamentals/oop_fundamentals.py
#
# Python Object-Oriented Programming Fundamentals
#
# Applying some concepts of classes and objects in Python programming.
# 
# Leonardo Franco de Godói
#
 
#------------------------------------------------------------------------------

# First parent class
class Box:
    
    # Init Method
    def __init__(self, Width, Height, Length):
        self.Width = Width
        self.Height = Height
        self.Length = Length

    # Method for resizing box dimensions
    def resize(self,w,h,l):
        self.Width = w * self.Width
        self.Height = h * self.Height
        self.Length = l * self.Length

    # Method for retrieving box volume
    def getVolume(self):
        Volume = self.Width * self.Height * self.Length
        return Volume
      
# Second parent class
class Data:  
    # Attributes
    trackingCode = 0
    zipCode = 0
    weight = 0

# Child class
class Delivery(Box, Data):
    # Attributes
    maxHeight = 0
    maxWeight = 0

    # Method for checking delivery parameters
    def check(self):
        if self.Height <= self.maxHeight and self.Weight <= self.maxWeight:
            print("All requirements properly met.\n")
        elif self.Height <= self.maxHeight: 
            print("Weight above permitted.\n")
        elif self.Weight <= self.maxWeight:
            print("Height above permitted.\n")
        else: 
            print("Height and Weight above permitted.\n")

# Instantiating a delivery object
myDelivery = Delivery(14.5, 8.5,9.0);

# Setting delivery attributes
myDelivery.ZipCode = 13822428;
myDelivery.TrackingCode = 102056800;
myDelivery.Weight = 3.42;

# Computing original delivery volume
print("Volume of box ", myDelivery.TrackingCode, " before resizing : ", myDelivery.getVolume(), "\n");
myDelivery.check()

# Resizing delivery
myDelivery.resize(0.5, 0.5, 2.0)

# Computing new delivery volume
print("Volume of box ", myDelivery.TrackingCode, " before resizing : ", myDelivery.getVolume(), "\n")
myDelivery.check()

File: c++/fundamentals/oop_fundamentals.cpp
/******************************************************************************

C++ Object-Oriented Programming Fundamentals

Reference for using classes and objects in C++.

Leonardo Franco de Godói

*******************************************************************************/

#include <iostream>

using namespace std;

// Parent class #1
class Box {
    // Class attributes
    public:
        double Length; 
        double Width; 
        double Height;
        
    // Class methods  
    public:
        void resize(float l, float w, float h) {
            Length = l * Length;
            Width = w * Width;
            Height = h * Height;
        }
        double getVolume() {
            double Volume = Height * Length * Width;
            return Volume;
        }
};

// Parent class #2
class Data {
    // Class attributes
    public:
        int ZipCode;
        int TrackingCode;
        float Weight;
};

// Derived Class (in order to show how inheritance works)
class Delivery: public Box, public Data {
    // Class attributes
    private:
        double maxHeight;
        float maxWeight;
        
    // Class methods
    public:
        void check() {
            if(Height <= maxHeight && Weight <= maxWeight) {
                cout << "All requirements properly met." << endl;
            }
            else if(Height <= maxHeight) {
                cout << "Weight above permitted." << endl;
            }
            else if(Weight <= maxWeight) {
                cout << "Height above permitted." << endl;
            }
            else {
                cout << "Height and Weight above permitted." << endl;
            }
        }
};

// Main function
int main()
{
    // Instantiating a delivery object
    Delivery myDelivery;
    
    // Setting delivery attributes
    myDelivery.Length = 14.5;
    myDelivery.Width = 8.5;
    myDelivery.Height = 9.0;
    myDelivery.ZipCode = 13822428;
    myDelivery.TrackingCode = 102056800;
    myDelivery.Weight = 3.42;
    
    // Computing original delivery volume
    cout << "Volume of box " << myDelivery.TrackingCode << " before resizing : " << myDelivery.getVolume() << endl;
    myDelivery.check();
    
    // Resizing delivery
    myDelivery.resize(0.5, 0.5, 2.0);
    
    // Computing new delivery volume
    cout << "Volume of box " <<  myDelivery.TrackingCode  << " before resizing : " << myDelivery.getVolume() << endl;
    myDelivery.check();

    return 0;
}